### PR TITLE
Aggregated optimization changes

### PR DESCRIPTION
**Pull request by monrostar**
Related issue: https://github.com/vendure-ecommerce/vendure/issues/2738

### CODE REVIEW

**Comment by monrostar** [comment]
@michaelbromley 
I created this PR to pull all the changes together and refactor them to avoid conflicts. 

In this PR I will optimize queries inside TransactionalConnection, it turned out to be difficult because there are assets inside collections that don't have reverse relation. We need to figure something out, maybe this relation can be found through metadata

**Comment by monrostar** [comment]
image
@michaelbromley 
We are just a little bit away from the final optimization of the most important methods

**Comment by tianyingchun** [comment]
how progress of this PR :)

**Comment by monrostar** [comment]
When i will be available

**Comment by monrostar** [comment]
PR is ready 🚀

File: packages/core/src/service/helpers/utils/tree-relations-qb-joiner.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
+/**
+ * Verifies if a relation has already been joined in a query builder to prevent duplicate joins.
+ * This method ensures query efficiency and correctness by maintaining unique joins within the query builder.
+ *
+ * @param {SelectQueryBuilder<T>} qb The query builder instance where the joins are being added.
+ * @param {string} alias The join alias to check for uniqueness. This alias is used to determine if the relation
+ *                       has already been joined to avoid adding duplicate join statements.
+ * @returns boolean Returns true if the relation has already been joined (based on the alias), false otherwise.
+ * @template T extends VendureEntity The entity type for which the query builder is configured.
+ */
```
Comment (michaelbromley) [inline packages/core/src/service/helpers/utils/tree-relations-qb-joiner.ts]: Looks like this doc block is out of place?

File: packages/core/src/entity/asset/orderable-asset.entity.ts [hunk 2]
```diff
@@ -24,7 +24,7 @@ export abstract class OrderableAsset extends VendureEntity implements Orderable
     assetId: ID;
 
     @Index()
-    @ManyToOne(type => Asset, { eager: true, onDelete: 'CASCADE' })
+    @ManyToOne(type => Asset,  { eager: true, onDelete: 'CASCADE' })
```
Comment (michaelbromley) [inline packages/core/src/entity/asset/orderable-asset.entity.ts]: extra space

File: packages/core/src/connection/transactional-connection.ts [hunk 3]
```diff
@@ -10,7 +11,9 @@ import {
     ObjectLiteral,
     ObjectType,
     Repository,
+    SelectQueryBuilder,
 } from 'typeorm';
+import { DataSource } from 'typeorm/data-source/DataSource';
```
Comment (michaelbromley) [inline packages/core/src/connection/transactional-connection.ts]: should this (and the `InjectDataSource` above) really be deep imports?

**Review by michaelbromley — COMMENTED** [review]
Excellent! Just a couple of small "tidy up" points. Thanks again for this incredible effort.

File: packages/core/src/service/helpers/utils/tree-relations-qb-joiner.ts (hunk 1 above)
Comment (monrostar) [reply packages/core/src/service/helpers/utils/tree-relations-qb-joiner.ts]: Yeah, I don't know how to do it here. If you have recommendations, I'd be glad to change it and maybe I need to reduce the amount of text...

File: packages/core/src/entity/asset/orderable-asset.entity.ts (hunk 2 above)
Comment (monrostar) [reply packages/core/src/entity/asset/orderable-asset.entity.ts]: ok, I will fix it

File: packages/core/src/connection/transactional-connection.ts (hunk 3 above)
Comment (monrostar) [reply packages/core/src/connection/transactional-connection.ts]: ok, I will fix it. this is from my IDE, I didn't noticed

Comment (monrostar) [reply packages/core/src/connection/transactional-connection.ts]: done

> should this (and the `InjectDataSource` above) really be deep imports?

File: packages/core/src/entity/asset/orderable-asset.entity.ts (hunk 2 above)
Comment (monrostar) [reply packages/core/src/entity/asset/orderable-asset.entity.ts]: > extra space

done

File: packages/core/src/service/helpers/utils/tree-relations-qb-joiner.ts (hunk 1 above)
Comment (monrostar) [reply packages/core/src/service/helpers/utils/tree-relations-qb-joiner.ts]: > Looks like this doc block is out of place?

I deleted the documentation that was accidentally copied and updated the documentation I was adding for the method. Check it out, please.

**Pull request merged**